### ISSUE TITLE
Function parameter hints

### ISSUE THREAD

**Issue by Mr-SD** [dragos/dragos-vscode-scala#39]
In Intellij you can have function parameters pop up which is super helpful:

![vscode-parameter-hints](https://cloud.githubusercontent.com/assets/3237827/25308282/0eb8bae6-2787-11e7-8825-34e3e6c31baf.gif)

I think it'd be nice to have that feature in vs code as well:
![intellij-parameter-hints](https://cloud.githubusercontent.com/assets/3237827/25308283/0ed36116-2787-11e7-8756-202551f56c4b.gif)

**Comment by dragos** [comment]
It's a great suggestion, but I'm not sure Ensime has such a feature, and currently we're relying on Ensime for all functionality.

If you're motivated in contributing this, I'd be happy to help you get started.

**Comment by fommil** [comment]
We certainly get parameter names in completions so I think this is possible. I'm not sure if the current API is enough, it may need another query.

**Comment by fommil** [comment]
I wonder if symbol-at-point provides this...

**Comment by dragos** [comment]
I'm wondering as well. The challenge is that the symbol at point may be something completely different from the actual call:

```
Car(ident<CARET>, )
```

Symbol-at-point will return `ident`, but we'd need `Car.apply`. In Eclipse we cut some corners and provided these for the initial completion call through Ctrl-Space. So that would work *outside* the parenthesis, but not inside.

There's a bunch of helper methods working on parse trees for getting enclosing trees, for instance [this one](https://github.com/scala-ide/scala-ide/blob/master/org.scala-ide.sdt.core/src/org/scalaide/core/compiler/CompilerApiExtensions.scala#L21) that might be used to find out the position of the enclosing call. But the main problem is that Ensime does not give access to the AST or compiler even when used as a library, if I understand correctly, so this functionality needs to be added as an additional request.

**Comment by fommil** [comment]
yeah that's right. I actually really hoped that `scala.meta` would give us an ADT of the AST which would let us expose so much more... but they decided to do it differently. It would have to be a fresh effort and that is a pretty huge undertaking.

**Comment by jvican** [comment]
I think a more approachable way to bring this feature is to force the user to always have the caret in either a comma or a parenthesis (either opening or dangling). This way, the vscode integration can assume that the user means the enclosing call and ask for the symbol at point of that concrete name.
  
I don't know how vscode integration handles positions, but it should be possible to pass in the position of the enclosing call instead of the position that the caret is located at. After all, clients have full control over these.

**Comment by jvican** [comment]
By the way, if someone has the vim integration installed for vscode, they can get something similar to tooltips by typing the empty constructor and then typing `g + h` to get the information about the type error they'll get in case that empty constructor doesn't exist (which is almost always).
